### DISCUSSION TITLE
Remote cache: Configurable upload timeout

### DISCUSSION THREAD

**Discussion by EPMatt** [vercel/turborepo#9472] [Ideas]
### Goals

1. Add the option to customize the artifacts upload timeout for the remote caching feature, to a different value rather than the statically defined 200 seconds window:

https://github.com/vercel/turborepo/blob/7fb5b21fe3034c29affef8e35a8c90ee75c0cb67/crates/turborepo-cache/src/http.rs#L333-L339

### Non-goals

1. Timeout for local tasks: this request is only concerned about the **upload timeout** for the remote caching feature.
2. Increasing the default upload timeout (200 seconds): instead, provide a way to customize the default.

### Background

When using a custom self-hosted server as remote cache, it's possible that the available bandwidth to the custom instance is limited. Moreover, the artifact size can be large enough for preventing the element being uploaded within the statically-configured upload timeout.

Having the possibility to override the default timeout would greatly improve the flexibility of the remote caching feature, allowing to set up custom servers to cache larger artifacts (100MB+) that require minutes to be uploaded.

For reference, users have reported turbo timing out while uploading artifacts on a custom remote cache server:
https://github.com/ducktors/turborepo-remote-cache/discussions/407

### Proposal

The remote timeout could be configured with an environment variable `TURBO_API_UPLOAD_TIMEOUT` similar to how the remote cache server can be set up with `TURBO_TEAM`, `TURBO_TOKEN` and `TURBO_API` variables.

Happy to work on the proposed feature if this is approved! 🚀

**Comment by anthonyshew** [comment]
As it turns out, we wrote the functionality for this 6 months ago and never finished it. 😁

Thanks for the reminder. Merged in https://github.com/vercel/turborepo/pull/9491.

**Reply by EPMatt** [reply]
That's great, thanks!
